### ISSUE TITLE
Retrieve data by Query 

### ISSUE THREAD

**Issue by farfilli** [rmcanany/SolidEdgeHousekeeper#161]
Example: Query my SQL Server DB to retrieve the part description and use it in the Title property.

I would add a settings panel where insert the Server connection string.

TBD Somewhere store the common queries. Structure them to use input parameters same as propertyTask

A query could retrieve more than a value then decide how to store and use them.

**Comment by rmcanany** [comment]
You're outta my league in the SQL department, but I can help with the save/retrieve part.

**Comment by farfilli** [comment]
If you have time and add a panel with two texboxes as a start, one for the connection string and one for the query.

Provide an example on how to use the text of those two text boxes would be enough as a start.

**Comment by rmcanany** [comment]
Can do.

**Comment by farfilli** [comment]
Also, I need to know how to retrieve their values to be used in the code, have you named some properties?
The query will contain input parameter in the usual format **%{System.Title}**

**Comment by rmcanany** [comment]
Added public properties `ServerConnectionString` and `ServerQuery`.  They should be saved and restored correctly now.

Not sure how the returned values from the server would be stored/identified/used.  I was hoping that would be your problem. :smiley:

**Comment by farfilli** [comment]
Best would be to let users use a formula like %{server1.query1.result1} that will produce a result value

Ideally we should be able to store more than one query, and perhaps even more than one server

I will work on the query side next week

**Comment by farfilli** [comment]
I have done a very simple implementation of the query to start coding; it's triggered by a console.writeline after a property edit on the ListViewFiles here: https://github.com/rmcanany/SolidEdgeHousekeeper/commit/4a05535da2623f01b523be9c6ae2ffa7f1a254ed#diff-a640283a9313a42d0466bb813b6d68afb8fefb020c89d3c471ae11533b5fe122R3069

The query that contains a property retrieve text needs to be implemented, I have just inserted a replacement text for testing purposes here: https://github.com/rmcanany/SolidEdgeHousekeeper/commit/4a05535da2623f01b523be9c6ae2ffa7f1a254ed#diff-a640283a9313a42d0466bb813b6d68afb8fefb020c89d3c471ae11533b5fe122R3256

Last but not least, I have added an example access db in the project folder, for some reason it creates a My Project folder when in the binary; perhaps you know a way to provide this example file in a better folder, fell free to move it as you think is better to distribute it.

To open the MDB you need the MS OLE Components, its something to install, I know it's not ideal, but let me know what do you think. I will add a SQL connection example, maybe with an online example DB

**Comment by rmcanany** [comment]
Sounds good.  Honestly, it's a bit over my head.  Let me know if you need anything from me.

**Comment by rmcanany** [comment]
OK, I'll take a look at that.

I ran it with the example you provided.  It didn't complain and it remembered a property edit I made by double-clicking a cell in the ListView.  

I noticed it added the following lines to `Housekeeper.vbproj`.  It looks like a machine-specific location, but it calls it a `<HintPath>`, so maybe it's OK.

```
    <Reference Include="System">
      <HintPath>..\..\..\..\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.7.2\System.dll</HintPath>
    </Reference>
```

**Comment by farfilli** [comment]
I have implemented the Query in SubstitutePropertyFormula; the syntax to use it is **%{System.Query1}** and it returns the first value of the query execution. The query itself may contain properties that are similarly correctly converted.

Here is an example where I retrieved the Subject value from the Query and wrote in the property
![image](https://github.com/user-attachments/assets/91cb028c-2de1-4301-9b3a-cc9ea19b5fc9)
![image](https://github.com/user-attachments/assets/c51fb136-f474-486c-abf0-1e5d1c452d9b)

**Comment by farfilli** [comment]
Do you think an online testing SQL database would be useful or the access DB is enough?
I found [this ](https://freeasphosting.net/free-sql-server-online.html)free service, never used it.

**Comment by rmcanany** [comment]
To me the Access DB is enough, but I'll let you decide.  

What do you think about adding a new PropertySetName `Server`?  Wouldn't that make it easier to parse the property formula?

**Comment by farfilli** [comment]
> To me the Access DB is enough, but I'll let you decide.
> 
> What do you think about adding a new PropertySetName `Server`? Wouldn't that make it easier to parse the property formula?

Yes it could be useful to differentiate what is a query and what not.

**Comment by rmcanany** [comment]
I'm working on turning TemplatePropertiesDict into a class.  It's a lot cleaner for sure with better control.  I've got Enums and everything!  I'll add `Server` to `PropertySetNameConstants`.

If the query were a Task, wouldn't that mess up the property filter and listview properties?  Those happen before processing, right?

**Comment by farfilli** [comment]
Let's leave the query thing separate from tasks by now, I will think about it another time

**Comment by farfilli** [comment]
Syntax improved with #171 
I have implemented %{Server.Query|Rx}

**Comment by rmcanany** [comment]
I like that.  If you don't get to it, I'll add `Server` to `PropertySetNameConstants` when I get back.

**Comment by rmcanany** [comment]
This is done, right?

**Comment by farfilli** [comment]
Yes it is
